### ISSUE TITLE
Errors in OpenCL mode

### ISSUE THREAD

**Issue by noelchalmers** [CEED/Laghos#20]
Hi I'm encountering errors with Laghos using OCCA 1.0 in OpenCL mode.  Command 
```sh
./laghos -p 1 -m data/square01_quad.mesh -rs 3 -tf 0.8 -no-vis -d "mode: 'OpenCL', device_id: 0, platform_id: 0 "
```
Output on a Titan V: 
```sh 

       __                __                 
      / /   ____  ____  / /_  ____  _____   
     / /   / __ `/ __ `/ __ \/ __ \/ ___/ 
    / /___/ /_/ / /_/ / / / / /_/ (__  )    
   /_____/\__,_/\__, /_/ /_/\____/____/  
               /____/                       

Options used:
   --mesh data/square01_quad.mesh
   --refine-serial 3
   --refine-parallel 0
   --problem 1
   --order-thermo 1
   --order-kinematic 2
   --ode-solver 4
   --t-final 0.8
   --cfl 0.1
   --no-visualization
   --no-visit
   --visualization-steps 5
   --outputfilename Laghos
   --device-info mode: 'OpenCL', device_id: 0, platform_id: 0 
   --occa-config 
   --no-occa-verbose
   --no-timings
Number of kinematic (position, velocity) dofs: 2178
Number of specific internal energy dofs: 1024
Repeating step 1
Repeating step 2
Repeating step 2
pure virtual method called
terminate called without an active exception
[paripp1:85275] *** Process received signal ***
[paripp1:85275] Signal: Aborted (6)
[paripp1:85275] Signal code:  (-6)
[paripp1:85275] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7fa889019390]
[paripp1:85275] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38)[0x7fa888c73428]
[paripp1:85275] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7fa888c7502a]
[paripp1:85275] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x16d)[0x7fa8897d384d]
[paripp1:85275] [ 4] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8d6b6)[0x7fa8897d16b6]
[paripp1:85275] [ 5] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8d701)[0x7fa8897d1701]
[paripp1:85275] [ 6] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8e23f)[0x7fa8897d223f]
[paripp1:85275] [ 7] /home/nchalmer/laghos/occa/lib/libocca.so(_ZN4occa6device20removeDHandleRefFromERPNS_8device_vE+0x21)[0x7fa88a142721]
[paripp1:85275] [ 8] /home/nchalmer/laghos/occa/lib/libocca.so(_ZN4occa6kernel16removeKHandleRefEv+0x48)[0x7fa88a08f428]
[paripp1:85275] [ 9] ./laghos(_ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_N4occa6kernelEESt10_Select1stISA_ESt4lessIS5_ESaISA_EE8_M_eraseEPSt13_Rb_tree_nodeISA_E+0x1ce)[0x60ed9e]
[paripp1:85275] [10] /home/nchalmer/laghos/occa/lib/libocca.so(_ZN4occa8device_vD1Ev+0x33)[0x7fa88a143753]
[paripp1:85275] [11] /home/nchalmer/laghos/occa/lib/libocca.so(_ZN4occa6opencl6deviceD0Ev+0x9)[0x7fa88a1b4f89]
[paripp1:85275] [12] /home/nchalmer/laghos/occa/lib/libocca.so(_ZN4occa6device20removeDHandleRefFromERPNS_8device_vE+0x2f)[0x7fa88a14272f]
[paripp1:85275] [13] /home/nchalmer/laghos/occa/lib/libocca.so(_ZN4occa6memory16removeMHandleRefEv+0x4b)[0x7fa88a16cb7b]
[paripp1:85275] [14] ./laghos[0x420d82]
[paripp1:85275] [15] ./laghos[0x419b4f]
[paripp1:85275] [16] ./laghos[0x411b61]
[paripp1:85275] [17] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7fa888c5e830]
[paripp1:85275] [18] ./laghos[0x4149a9]
[paripp1:85275] *** End of error message ***
Aborted (core dumped)
```

 and output on a Radeon Frontier:
```sh
       __                __                 
      / /   ____  ____  / /_  ____  _____   
     / /   / __ `/ __ `/ __ \/ __ \/ ___/ 
    / /___/ /_/ / /_/ / / / / /_/ (__  )    
   /_____/\__,_/\__, /_/ /_/\____/____/  
               /____/                       

Options used:
   --mesh data/square01_quad.mesh
   --refine-serial 3
   --refine-parallel 0
   --problem 1
   --order-thermo 1
   --order-kinematic 2
   --ode-solver 4
   --t-final 0.8
   --cfl 0.1
   --no-visualization
   --no-visit
   --visualization-steps 5
   --outputfilename Laghos
   --device-info mode: 'OpenCL', device_id: 0, platform_id: 0 
   --occa-config 
   --no-occa-verbose
   --no-timings
Number of kinematic (position, velocity) dofs: 2178
Number of specific internal energy dofs: 1024

---[ Error ]------------------------------------------------
    File     : /home/noel/laghos/occa/src/modes/opencl/memory.cpp
    Function : addOffset
    Line     : 71
    Message  : Device: clCreateSubBuffer
    Error   : OpenCL Error [ -13 ]: CL_MISALIGNED_SUB_BUFFER_OFFSET
    Stack    :
      8 occa::opencl::error(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
      7 occa::opencl::memory::addOffset(long, bool&)
      6 occa::memory::slice(long, long) const
      5 occa::memory::operator+(long) const
      4 ./laghos()
      3 ./laghos()
      2 /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)
      1 ./laghos()
============================================================
terminate called after throwing an instance of 'int'
Aborted (core dumped)

```

Any suggestions would be helpful.

**Comment by dmed256** [comment]
I haven't ported over Laghos+OCCA to OCCA 1.0

However, MFEM+OCCA doesn't support OpenCL due to
- ~~https://github.com/libocca/occa/issues/136~~
- https://github.com/libocca/occa/issues/166

Not sure if https://github.com/libocca/occa/issues/136 is enough, will have to check later

**Comment by noelchalmers** [comment]
I've also tried this in my fork of occa which fixed libocca/occa#166. At least for the AMD card, the error occurs when using an offset to make an OpenCL sub-buffer. I think this could potentially have a work around. Not sure why the Nvidia card has a different error, however.

**Comment by dmed256** [comment]
I'm going to close this since it's more of an OCCA bug than Laghos bug. Although it's not a high priority at the moment, once https://github.com/libocca/occa/issues/166 is resolved Laghos should work ok with OpenCL.
